### PR TITLE
clean up UI display of recent Chef logs on failure

### PR DESCRIPTION
**Pull request by aspiers**
When `chef-client` fails, the most recent lines of the Chef logs are surfaced in the Crowbar proposal UI.  However they are not wrapped in anything, so the line breaks gets ignored, creating an unholy mess.  Not sure how this was tolerated for so long, but let's fix it.

Also a couple of refactorings to clean up the relevant method.

### CODE REVIEW

**Comment by susecloudcommitpolice** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please limit commit subject line to 50 characters.<br />
fda8ea9c187daaf3a3bbd1c8da4ffa3c17a06981</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by aspiers** [comment]
@nicolasbock Will do; I would have done that instantly but it took me a few moments to understand the cryptic variable names.

**Comment by nicolasbock** [comment]
I am not surprised ;)

**Comment by aspiers** [comment]
@nicolasbock Fixed. I don't agree with the remaining Hound warning.

**Comment by aspiers** [comment]
Added a PoC attempt to get a nicer background colour.  Pretty sure this is wrong somehow; I've never made changes which could affect the branding CSS.

**Comment by jsuchome** [comment]
@aspiers do you have some failure example, so we can see the differences? (not necessary screenshot, just the way to reproduce some error you're testing)

BTW, adding @skazi0 , he's far better than me in this areas

**Comment by aspiers** [comment]
@jsuchome Any proposal failure - they are not hard to find, just check the Jenkins Green boards ;-) But yes I should include a screenshot here.

**Comment by jsuchome** [comment]
So this is how it looks for me http://imagebin.suse.de/2107/img ... but I have not been able to use that `state-danger-pre-bg` variable, `rake assets:precompile` fails with Undefined variable: "$state-danger-pre-bg". ... which is probably caused my missing knowledge of how to correctly apply your patch...

**Comment by aspiers** [comment]
@jsuchome Thanks for testing!  I did not test the SASS changes yet, so it's easily possible I screwed up somehow.  But interesting that you did not get the same white background for the Chef logs which I got (that's what the variable is supposed to fix).

Yes, you can see from the screenshot that there is an unwanted extra blank line in between every line of the Chef logs, but I have no clue where these extra `<br>` elements are coming from :-/

File: crowbar_framework/vendor/assets/stylesheets/variables.scss [hunk 1]
```diff
@@ -340,6 +340,7 @@ $state-warning-border: darken(adjust-hue($state-warning-bg, -10), 5%);
 
 $state-danger-text: #a94442;
 $state-danger-bg: #f2dede;
+$state-danger-pre-bg: lighten($state-danger-bg, 5%);
```
Comment (skazi0) [inline crowbar_framework/vendor/assets/stylesheets/variables.scss]: Move this to `bootstrap/_variables.scss`. `variables.scss` doesn't seem to be included anywhere.

File: crowbar_framework/app/models/service_object.rb [hunk 2]
```diff
@@ -1814,23 +1814,29 @@ def handle_validation_errors
...
+        # If we found a delimiter in the last (say) 10 lines, we don't need
+        # to show all of the last 50.
+        if (last_delimiter_line > 0) && (line_count - last_delimiter_line) < 50
+          last_delimiter_line - 3
+        else
+          line_count - 50
+        end
+      logged_lines = f.readlines[starting_line..line_count]
+      "Most recent logged lines from the Chef run: \n\n<pre>" +
+        logged_lines.join(" ") + "</pre>"
```
Comment (skazi0) [inline crowbar_framework/app/models/service_object.rb]: Why not join with empty string? The space appears as an indentation starting from second line.

Comment (aspiers) [reply crowbar_framework/app/models/service_object.rb]: Oops, good point :) Fixed.

File: crowbar_framework/vendor/assets/stylesheets/variables.scss (hunk 1 above)
Comment (aspiers) [reply crowbar_framework/vendor/assets/stylesheets/variables.scss]: Thanks, fixed.

**Comment by aspiers** [comment]
> I think the problem with <BR> comes from: https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/app/views/shared/_flashes.html.haml#L6
>
> See: https://apidock.com/rails/ActionView/Helpers/TextHelper/simple_format

Good catch, that's it!  Any ideas how to cleanly fix it to stop doing that in this particular case?

**Comment by skazi0** [comment]
@aspiers I don't have a ready solution. I will try to find something tomorrow.

**Comment by aspiers** [comment]
@skazi0 Thanks, really appreciate your help :)

**Comment by skazi0** [comment]
@aspiers the best solution I found is this: 

    = simple_format(message).tr("\n", '').html_safe

`simple_format` converts newlines to `<br>` or '<p>` but doesn't remove the newlines. In normal HTML these should be ignored and rendering is based on the above tags. We can safely remove the newlines and the rendering should be correct. `html_safe` flag set by `simple_format` is lost after `tr` so we need to mark the string as safe again.

Alternative is more hacky, css-only, not-sure-how-portable: `pre br { display: none; }` (you should probably add some class to this `<pre>` and use that in css to not break any other views).

**Comment by aspiers** [comment]
Jacek Tomasiak <notifications@github.com> wrote:
>@aspiers the best solution I found is this:
>
>    = simple_format(message).tr("\n", '').html_safe
>
>`simple_format` converts newlines to `<br>` or '<p>` but doesn't remove the newlines. In normal HTML these should be ignored and rendering is based on the above tags. We can safely remove the newlines and the rendering should be correct. `html_safe` flag set by `simple_format` is lost after `tr` so we need to mark the string as safe again.

That would work, but since this is done in
app/views/shared/_flashes.html.haml, how would we actually implement
that for flashes which contain Chef logs or other <pre> blocks,
without breaking all the other flashes out there?

>Alternative is more hacky, css-only, not-sure-how-portable: `pre br { display: none; }` (you should probably add some class to this `<pre>` and use that in css to not break any other views).

Ooooh that's evil! ... I like it ;-)  By far the easiest solution.  If
you look at the PR you'll see that the CSS for this <pre> is already
wrapped inside an .alert { } block, so it wouldn't break anything
else.  I'll try it.

File: crowbar_framework/app/assets/stylesheets/content/alert.scss [hunk 3]
```diff
@@ -7,4 +7,14 @@
   p:last-child {
     margin-bottom: 0
   }
+
+  pre {
+    background-color: $state-danger-pre-bg;
+    overflow: scroll;
+    max-height: 20em;
+
+    br {
```
[the remark this answers is not in the document]
Comment (aspiers) [reply crowbar_framework/app/assets/stylesheets/content/alert.scss]: It looks like here Hound is being more aggressive than [the default behaviour of `scss-lint`](https://github.com/brigade/scss-lint/blob/3c1aeb4d7a5a5aba0a7aa1fc2e8fcd929026df36/lib/scss_lint/linter/README.md#selectordepth); I think this should be ignored.

**Comment by aspiers** [comment]
Here's roughly what it looks like now, although I hacked the CSS into Chrome manually rather than going through the whole asset compilation dance, so the background colour for the `<pre>` might be wrong still.

chef-failures

**Comment by aspiers** [comment]
Still not sure why the first line is outdented by 1 column.

**Comment by nicolasbock** [comment]
But this does look a whole lot better than before!

**Comment by skazi0** [comment]
@aspiers re outdent: have you re-applied the proposal after removing space from join? the rendered outputs are stored in proposal so only css stuff gets updated without re-applying.

**Comment by skazi0** [comment]
@aspiers re my first solution and breaking other flashes: I don't think it changes anything besides fixing the empty lines in `<pre>`.

My thinking:

```
message = "first line\nsecond line"
simple_format(message) == "first line\n<br />second line"
```
When this is rendered as HTML the `\n` is ignored by the browser. The only exception is in `<pre>` or similar (e.g. css: `white-space: pre`). I don't see how removing the newlines from `simple_format` output could break anything anywhere.

**Comment by aspiers** [comment]
> @aspiers re outdent: have you re-applied the proposal after removing space from join? the rendered outputs are stored in proposal so only css stuff gets updated without re-applying.

Yes this was a fresh test. The problem is somewhere else.

**Comment by skazi0** [comment]
@aspiers re outdent again, take a closer look, it's not an outdent but a missing indent on the first line. Spaces are there for padding the line numbers.

**Comment by aspiers** [comment]
@skazi0 Not sure I follow? But no outdent here, weirdly! - http://crowbar.vk3.cloud.suse.de/crowbar/neutron/1.0/proposals/default

**Comment by aspiers** [comment]
Neutron failed in the gate (totally unrelated of course), giving us a nice opportunity to examine the results of this, and it looks good. So I think this is ready to go.

**Pull request merged**